src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Integer/DenseIntegralFrozenDictionary.cs

```diff
@@ -0,0 +1,239 @@
+
+            if (typeof(TKey) == typeof(byte) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(byte)))
+                return TryCreate<TKey, byte, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(sbyte) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(sbyte)))
+                return TryCreate<TKey, sbyte, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(ushort) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(ushort)))
+                return TryCreate<TKey, ushort, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(short) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(short)))
+                return TryCreate<TKey, short, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(int) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(int)))
+                return TryCreate<TKey, int, TValue>(source, out result);
+
+            result = null;
+            return false;
+        }
+
+        private static bool TryCreate<TKey, TKeyUnderlying, TValue>(Dictionary<TKey, TValue> source, [NotNullWhen(true)] out FrozenDictionary<TKey, TValue>? result)
+            where TKey : notnull
+            where TKeyUnderlying : unmanaged, IBinaryInteger<TKeyUnderlying>
+        {
+            // Start enumerating the dictionary to ensure it has at least one element.
+            Dictionary<TKey, TValue>.Enumerator e = source.GetEnumerator();
+            if (e.MoveNext())
+            {
+                // Get that element and treat it as the min and max. Then continue enumerating the remainder
+                // of the dictionary to count the number of elements and track the full min and max.
+                int count = 1;
+                int min = int.CreateTruncating((TKeyUnderlying)(object)e.Current.Key);
+                int max = min;
+                while (e.MoveNext())
+                {
+                    count++;
+                    int key = int.CreateTruncating((TKeyUnderlying)(object)e.Current.Key);
+                    if (key < min)
+                    {
+                        min = key;
+                    }
+                    else if (key > max)
+                    {
+                        max = key;
+                    }
+                }
+
+                // Based on the min and max, determine the spread. If the range fits within a non-negative Int32
+                // and the ratio of the number of elements in the dictionary to the length is within the allowed
+                // threshold, create the new dictionary.
+                long length = (long)max - min + 1;
+                Debug.Assert(length > 0);
+                if (length <= int.MaxValue &&
+                    (double)count / length >= CountToLengthRatio)
```

Comment:
subjective : this method has a lot of nested `if` and `loop` blocks. We could reduce that by bailing out early.

```cs
if (length > int.MaxValue || (double)count / length < CountToLengthRatio)
{
    result = null;
    return false;
}
```

Reply:
Now this is the only branch that can return false because the dictionary can't be empty as I mentioned before. I will feel better if the block can reduce one more indent.